File: src/main/java/com/milaboratory/mixcr/basictypes/MultiAlignmentHelper.java
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.basictypes;

import com.milaboratory.core.Range;
import com.milaboratory.core.alignment.Alignment;
import com.milaboratory.core.mutations.MutationType;
import com.milaboratory.core.mutations.Mutations;
import com.milaboratory.core.sequence.Alphabet;
import com.milaboratory.core.sequence.Sequence;
import com.milaboratory.core.sequence.SequenceQuality;
import com.milaboratory.util.BitArray;
import com.milaboratory.util.IntArrayList;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.milaboratory.core.mutations.Mutation.RAW_MUTATION_TYPE_DELETION;
import static com.milaboratory.core.mutations.Mutation.RAW_MUTATION_TYPE_SUBSTITUTION;

public class MultiAlignmentHelper {
    // subject / queries nomenclature seems to be swapped here...
    // subject here corresponds to sequence2 from alignments (so, the query sequence)
    // queries here corresponds to sequence1 from alignments (so, the reference sequence)

    int minimalPositionWidth = 0;
    final String subject;
    final String[] queries;
    final int[] subjectPositions;
    final int[][] queryPositions;
    final BitArray[] match;
    final List<String> annotationStrings = new ArrayList<>();
    final List<String> annotationStringTitles = new ArrayList<>();

    String subjectLeftTitle;
    final String[] queryLeftTitles;

    String subjectRightTitle;
    final String[] queryRightTitles;

    private MultiAlignmentHelper(String subject, String[] queries, int[] subjectPositions,
                                 int[][] queryPositions, BitArray[] match,
                                 String subjectLeftTitle, String[] queryLeftTitles,
                                 String subjectRightTitle, String[] queryRightTitles) {
        this.subject = subject;
        this.queries = queries;
        this.subjectPositions = subjectPositions;
        this.queryPositions = queryPositions;
        this.match = match;
        this.subjectLeftTitle = subjectLeftTitle;
        this.queryLeftTitles = queryLeftTitles;
        this.subjectRightTitle = subjectRightTitle;
        this.queryRightTitles = queryRightTitles;
    }

    private MultiAlignmentHelper(String subject, String[] queries, int[] subjectPositions, int[][] queryPositions,
                                 BitArray[] match) {
        this(subject, queries, subjectPositions, queryPositions, match, "", new String[queries.length],
                "", new String[queries.length]);
    }

    public String getSubject() {
        return subject;
    }

    public String getQuery(int idx) {
        return queries[idx];
    }

    public int[] getSubjectPositions() {
        return subjectPositions;
    }

    public int[][] getQueryPositions() {
        return queryPositions;
    }

    public BitArray[] getMatch() {
        return match;
    }

    public String getSubjectLeftTitle() {
        return subjectLeftTitle;
    }

    public String getSubjectRightTitle() {
        return subjectRightTitle;
    }

    public String getQueryLeftTitle(int i) {
        return queryLeftTitles[i];
    }

    public String getQueryRightTitle(int i) {
        return queryRightTitles[i];
    }

    public int getActualPositionWidth() {
        int ret = ("" + getSubjectFrom()).length();
        for (int i = 0; i < queries.length; i++)
            ret = Math.max(ret, ("" + getQueryFrom(i)).length());
        return ret;
    }

    public void setMinimalPositionWidth(int minimalPositionWidth) {
        this.minimalPositionWidth = minimalPositionWidth;
    }

    public MultiAlignmentHelper setSubjectLeftTitle(String subjectLeftTitle) {
        this.subjectLeftTitle = subjectLeftTitle;
        return this;
    }

    public MultiAlignmentHelper addSubjectQuality(String title, SequenceQuality quality) {
        char[] chars = new char[size()];
        for (int i = 0; i < size(); ++i)
            chars[i] = subjectPositions[i] < 0 ? ' ' : simplifiedQuality(quality.value(subjectPositions[i]));
        addAnnotationString(title, new String(chars));
        return this;
    }

    private static char simplifiedQuality(int value) {
        value /= 5;
        if (value > 9)
            value = 9;
        return Integer.toString(value).charAt(0);
    }

    public MultiAlignmentHelper setSubjectRightTitle(String subjectRightTitle) {
        this.subjectRightTitle = subjectRightTitle;
        return this;
    }

    public MultiAlignmentHelper addAnnotationString(String title, String string) {
        if (string.length() != size())
            throw new IllegalArgumentException();
        annotationStrings.add(string);
        annotationStringTitles.add(title);
        return this;
    }

    public MultiAlignmentHelper setQueryLeftTitle(int id, String queryLeftTitle) {
        this.queryLeftTitles[id] = queryLeftTitle;
        return this;
    }

    public MultiAlignmentHelper setQueryRightTitle(int id, String queryRightTitle) {
        this.queryRightTitles[id] = queryRightTitle;
        return this;
    }

    public int getSubjectPositionAt(int position) {
        return subjectPositions[position];
    }

    public int subjectToAlignmentPosition(int subjectPosition) {
        for (int i = 0; i < subjectPositions.length; i++)
            if (subjectPositions[i] == subjectPosition)
                return i;
        return -1;
    }

    public int getQueryPositionAt(int index, int position) {
        return queryPositions[index][position];
    }

    public int getAbsSubjectPositionAt(int position) {
        return aabs(subjectPositions[position]);
    }

    public int getAbsQueryPositionAt(int index, int position) {
        return aabs(queryPositions[index][position]);
    }

    private static int aabs(int pos) {
        if (pos >= 0)
            return pos;
        if (pos == -1)
            return -1;
        return -2 - pos;
    }

    public int getSubjectFrom() {
        return getFirstPosition(subjectPositions);
    }

    public int getSubjectTo() {
        return getLastPosition(subjectPositions);
    }

    public int getSubjectLength() {
        return 1 + getSubjectTo() - getSubjectFrom();
    }

    public int getQueryFrom(int index) {
        return getFirstPosition(queryPositions[index]);
    }

    public int getQueryTo(int index) {
        return getLastPosition(queryPositions[index]);
    }

    public String getAnnotationString(int i) {
        return annotationStrings.get(i);
    }

    public int size() {
        return subject.length();
    }

    public MultiAlignmentHelper getRange(int from, int to) {
        boolean[] queriesToExclude = new boolean[queries.length];
        int queriesCount = 0;
        for (int i = 0; i < queries.length; i++) {
            boolean exclude = true;
            for (int j = from; j < to; j++)
                if (queryPositions[i][j] != -1) {
                    exclude = false;
                    break;
                }
            queriesToExclude[i] = exclude;
            if (!exclude)
                queriesCount++;
        }

        String[] cQueries = new String[queriesCount];
        int[][] cQueryPositions = new int[queriesCount][];
        BitArray[] cMatch = new BitArray[queriesCount];
        String[] cQueryLeftTitles = new String[queriesCount];
        String[] cQueryRightTitles = new String[queriesCount];

        int j = 0;
        for (int i = 0; i < queries.length; i++) {
            if (queriesToExclude[i])
                continue;
            cQueries[j] = queries[i].substring(from, to);
            cQueryPositions[j] = Arrays.copyOfRange(queryPositions[i], from, to);
            cMatch[j] = match[i].getRange(from, to);
            cQueryLeftTitles[j] = queryLeftTitles[i];
            cQueryRightTitles[j] = queryRightTitles[i];
            j++;
        }

        MultiAlignmentHelper result = new MultiAlignmentHelper(subject.substring(from, to), cQueries,
                Arrays.copyOfRange(subjectPositions, from, to), cQueryPositions, cMatch,
                subjectLeftTitle, cQueryLeftTitles, subjectRightTitle, cQueryRightTitles);

        for (int i = 0; i < annotationStrings.size(); i++)
            result.addAnnotationString(annotationStringTitles.get(i),
                    annotationStrings.get(i).substring(from, to));
        return result;
    }

    public MultiAlignmentHelper[] split(int length) {
        return split(length, false);
    }

    public MultiAlignmentHelper[] split(int length, boolean eqPositionWidth) {
        MultiAlignmentHelper[] ret = new MultiAlignmentHelper[(size() + length - 1) / length];
        for (int i = 0; i < ret.length; ++i) {
            int pointer = i * length;
            int l = Math.min(length, size() - pointer);
            ret[i] = getRange(pointer, pointer + l);
        }
        if (eqPositionWidth)
            alignPositions(ret);
        return ret;
    }

    private static int getFirstPosition(int[] array) {
        for (int pos : array)
            if (pos >= 0)
                return pos;
        for (int pos : array)
            if (pos < -1)
                return -2 - pos;
        return -1;
    }

    private static int getLastPosition(int[] array) {
        for (int i = array.length - 1; i >= 0; i--)
            if (array[i] >= 0)
                return array[i];
        for (int i = array.length - 1; i >= 0; i--)
            if (array[i] < -1)
                return -2 - array[i];
        return -1;
    }

    private static int fixedWidthL(String[] strings) {
        return fixedWidthL(strings, 0);
    }

    private static int fixedWidthL(String[] strings, int minWidth) {
        int length = 0;
        for (String string : strings)
            length = Math.max(length, string.length());
        length = Math.max(length, minWidth);
        for (int i = 0; i < strings.length; i++)
            strings[i] = spaces(length - strings[i].length()) + strings[i];
        return length;
    }

    private static int fixedWidthR(String[] strings) {
        return fixedWidthR(strings, 0);
    }

    private static int fixedWidthR(String[] strings, int minWidth) {
        int length = 0;
        for (String string : strings)
            length = Math.max(length, string.length());
        length = Math.max(length, minWidth);
        for (int i = 0; i < strings.length; i++)
            strings[i] = strings[i] + spaces(length - strings[i].length());
        return length;
    }

    public static class Settings {
        public final boolean markMatchWithSpecialLetter;
        public final boolean lowerCaseMatch;
        public final boolean lowerCaseMismatch;
        public final char specialMatchChar;
        public final char outOfRangeChar;

        public Settings(boolean markMatchWithSpecialLetter, boolean lowerCaseMatch, boolean lowerCaseMismatch,
                        char specialMatchChar, char outOfRangeChar) {
            this.markMatchWithSpecialLetter = markMatchWithSpecialLetter;
            this.lowerCaseMatch = lowerCaseMatch;
            this.lowerCaseMismatch = lowerCaseMismatch;
            this.specialMatchChar = specialMatchChar;
            this.outOfRangeChar = outOfRangeChar;
        }
    }

    @Override
    public String toString() {
        int aCount = queries.length;
        int asSize = annotationStringTitles.size();

        String[] lines = new String[aCount + 1 + asSize];

        for (int i = 0; i < asSize; i++)
            lines[i] = "";

        lines[asSize] = "" + getSubjectFrom();
        for (int i = 0; i < aCount; i++)
            lines[i + 1 + asSize] = "" + getQueryFrom(i);

        int width = fixedWidthL(lines, minimalPositionWidth);

        for (int i = 0; i < asSize; i++)
            lines[i] = annotationStringTitles.get(i) + spaces(width + 1);

        lines[asSize] = (subjectLeftTitle == null ? "" : subjectLeftTitle) +
                " " + lines[asSize];

        for (int i = 0; i < aCount; i++)
            lines[i + 1 + asSize] = (queryLeftTitles[i] == null ? "" : queryLeftTitles[i]) +
                    " " + lines[i + 1 + asSize];

        width = fixedWidthL(lines);

        for (int i = 0; i < asSize; i++)
            lines[i] += " " + annotationStrings.get(i);
        lines[asSize] += " " + subject + " " + getSubjectTo();
        for (int i = 0; i < aCount; i++)
            lines[i + 1 + asSize] += " " + queries[i] + " " + getQueryTo(i);

        width = fixedWidthR(lines);

        lines[asSize] += " " + subjectRightTitle;
        for (int i = 0; i < aCount; i++)
            if (queryRightTitles[i] != null)
                lines[i + 1 + asSize] += " " + queryRightTitles[i];

        StringBuilder result = new StringBuilder();
        for (int i = 0; i < lines.length; i++) {
            if (i != 0)
                result.append("\n");
            result.append(lines[i]);
        }
        return result.toString();
    }

    public static final Settings DEFAULT_SETTINGS = new Settings(false, true, false, ' ', ' ');
    public static final Settings DOT_MATCH_SETTINGS = new Settings(true, true, false, '.', ' ');

    public static <S extends Sequence<S>> MultiAlignmentHelper build(Settings settings, Range subjectRange,
                                                                     Alignment<S>... alignments) {
        S subject = alignments[0].getSequence1();
        return build(settings, subjectRange, subject, alignments);
    }

    public static <S extends Sequence<S>> MultiAlignmentHelper build(Settings settings, Range subjectRange,
                                                                     S subject, Alignment<S>... alignments) {
        for (Alignment<S> alignment : alignments)
            if (!alignment.getSequence1().equals(subject))
                throw new IllegalArgumentException();

        int subjectPointer = subjectRange.getFrom();
        int subjectPointerTo = subjectRange.getTo();

        int aCount = alignments.length;
        int[] queryPointers = new int[aCount];
        int[] mutationPointers = new int[aCount];
        Mutations<S>[] mutations = new Mutations[aCount];
        List<Boolean>[] matches = new List[aCount];

        IntArrayList subjectPositions = new IntArrayList();
        IntArrayList[] queryPositions = new IntArrayList[aCount];

        StringBuilder subjectString = new StringBuilder();
        StringBuilder[] queryStrings = new StringBuilder[aCount];

        for (int i = 0; i < aCount; i++) {
            queryPointers[i] = alignments[i].getSequence2Range().getFrom();
            matches[i] = new ArrayList<>();
            mutations[i] = alignments[i].getAbsoluteMutations();
            queryPositions[i] = new IntArrayList();
            queryStrings[i] = new StringBuilder();
        }

        final Alphabet<S> alphabet = subject.getAlphabet();

        BitArray processed = new BitArray(aCount);
        while (true) {
            // Checking continue condition
            boolean doContinue = subjectPointer < subjectPointerTo;
            for (int i = 0; i < aCount; i++)
                doContinue |= mutationPointers[i] < mutations[i].size();
            if (!doContinue)
                break;

            processed.clearAll();

            // Processing out of range sequences
            for (int i = 0; i < aCount; i++)
                if (!alignments[i].getSequence1Range().contains(subjectPointer)
                        && !(alignments[i].getSequence1Range().containsBoundary(subjectPointer) &&
                        mutationPointers[i] != mutations[i].size())) {
                    queryStrings[i].append(settings.outOfRangeChar);
                    queryPositions[i].add(-1);
                    matches[i].add(false);
                    processed.set(i);
                }

            // Checking for insertions
            boolean insertion = false;
            for (int i = 0; i < aCount; i++)
                if (mutationPointers[i] < mutations[i].size() &&
                        mutations[i].getTypeByIndex(mutationPointers[i]) == MutationType.Insertion &&
                        mutations[i].getPositionByIndex(mutationPointers[i]) == subjectPointer) {
                    insertion = true;
                    queryStrings[i].append(mutations[i].getToAsSymbolByIndex(mutationPointers[i]));
                    queryPositions[i].add(queryPointers[i]++);
                    matches[i].add(false);
                    mutationPointers[i]++;
                    assert !processed.get(i);
                    processed.set(i);
                }

            if (insertion) { // In case on insertion in query sequence
                subjectString.append('-');
                subjectPositions.add(-2 - subjectPointer);

                for (int i = 0; i < aCount; i++) {
                    if (!processed.get(i)) {
                        queryStrings[i].append('-');
                        queryPositions[i].add(-2 - queryPointers[i]);
                        matches[i].add(false);
                    }
                }
            } else { // In other cases
                char subjectSymbol = subject.symbolAt(subjectPointer);
                subjectString.append(subjectSymbol);
                subjectPositions.add(subjectPointer);

                for (int i = 0; i < aCount; i++) {
                    if (processed.get(i))
                        continue;

                    Mutations<S> cMutations = mutations[i];
                    int cMutationPointer = mutationPointers[i];

                    boolean mutated = false;

                    if (cMutationPointer < cMutations.size()) {
                        int mutPosition = cMutations.getPositionByIndex(cMutationPointer);
                        assert mutPosition >= subjectPointer;
                        mutated = mutPosition == subjectPointer;
                    }

                    if (mutated) {
                        switch (cMutations.getRawTypeByIndex(cMutationPointer)) {
                            case RAW_MUTATION_TYPE_SUBSTITUTION:
                                char symbol = cMutations.getToAsSymbolByIndex(cMutationPointer);
                                queryStrings[i].append(settings.lowerCaseMismatch ?
                                        Character.toLowerCase(symbol) :
                                        symbol);
                                queryPositions[i].add(queryPointers[i]++);
                                matches[i].add(false);
                                break;
                            case RAW_MUTATION_TYPE_DELETION:
                                queryStrings[i].append('-');
                                queryPositions[i].add(-2 - queryPointers[i]);
                                matches[i].add(false);
                                break;
                            default:
                                assert false;
                        }
                        mutationPointers[i]++;
                    } else {
                        if (settings.markMatchWithSpecialLetter)
                            queryStrings[i].append(settings.specialMatchChar);
                        else
                            queryStrings[i].append(settings.lowerCaseMatch ? Character.toLowerCase(subjectSymbol) :
                                    subjectSymbol);
                        queryPositions[i].add(queryPointers[i]++);
                        matches[i].add(true);
                    }
                }
                subjectPointer++;
            }
        }

        int[][] queryPositionsArrays = new int[aCount][];
        BitArray[] matchesBAs = new BitArray[aCount];
        String[] queryStringsArray = new String[aCount];
        for (int i = 0; i < aCount; i++) {
            queryPositionsArrays[i] = queryPositions[i].toArray();
            matchesBAs[i] = new BitArray(matches[i]);
            queryStringsArray[i] = queryStrings[i].toString();
        }

        return new MultiAlignmentHelper(subjectString.toString(), queryStringsArray, subjectPositions.toArray(),
                queryPositionsArrays, matchesBAs);
    }

    public static void alignPositions(MultiAlignmentHelper[] helpers) {
        int maxPositionWidth = 0;
        for (MultiAlignmentHelper helper : helpers)
            maxPositionWidth = Math.max(maxPositionWidth, helper.getActualPositionWidth());
        for (MultiAlignmentHelper helper : helpers)
            helper.setMinimalPositionWidth(maxPositionWidth);
    }

    private static String spaces(int n) {
        char[] c = new char[n];
        Arrays.fill(c, ' ');
        return String.valueOf(c);
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/basictypes/MultiAlignmentHelper.kt
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.basictypes

import com.milaboratory.core.Range
import com.milaboratory.core.alignment.Alignment
import com.milaboratory.core.mutations.Mutation
import com.milaboratory.core.mutations.MutationType
import com.milaboratory.core.mutations.Mutations
import com.milaboratory.core.sequence.NucleotideSequence
import com.milaboratory.core.sequence.Sequence
import com.milaboratory.core.sequence.SequenceQuality
import com.milaboratory.util.BitArray
import com.milaboratory.util.IntArrayList
import io.repseq.core.SequencePartitioning
import kotlin.math.min

class MultiAlignmentHelper<S : Sequence<S>> private constructor(
    // subject / queries nomenclature seems to be swapped here...
    // subject here corresponds to sequence2 from alignments (so, the query sequence)
    // queries here corresponds to sequence1 from alignments (so, the reference sequence)
    val subject: SubjectLine<S>,
    val queries: Array<QueryLine>,
    val match: Array<BitArray>,
    val metaInfo: List<MetaInfoInput<S>>
) {


    fun getQuery(idx: Int): String = queries[idx].content

    val actualPositionWidth: Int
        get() = (queries.map { it.firstPosition.toString() } + subject.firstPosition.toString())
            .maxOf { it.length }

    fun subjectToAlignmentPosition(subjectPosition: Int): Int {
        for (i in subject.positions.indices) if (subject.positions[i] == subjectPosition) return i
        return -1
    }

    fun getAbsSubjectPositionAt(position: Int): Int = aabs(subject.positions[position])

    fun getAbsQueryPositionAt(index: Int, position: Int): Int = aabs(queries[index].positions[position])

    fun size(): Int = subject.content.length

    fun getRange(from: Int, to: Int): MultiAlignmentHelper<S> {
        val queriesToExclude = BooleanArray(queries.size)
        var queriesCount = 0
        for (i in queries.indices) {
            var exclude = true
            for (j in from until to) if (queries[i].positions[j] != -1) {
                exclude = false
                break
            }
            queriesToExclude[i] = exclude
            if (!exclude) queriesCount++
        }
        val indexMapping = IntArray(queriesCount)
        var j = 0
        for (i in queries.indices) {
            if (queriesToExclude[i]) continue
            indexMapping[j] = i
            j++
        }

        val cQueries: Array<QueryLine> = Array(queriesCount) {
            val origin = queries[indexMapping[it]]
            origin.subRange(from, to)
        }
        val cMatch: Array<BitArray> = Array(queriesCount) {
            match[indexMapping[it]].getRange(from, to)
        }
        return MultiAlignmentHelper(
            subject.subRange(from, to),
            cQueries,
            cMatch,
            metaInfo
        )
    }

    fun split(length: Int): Array<MultiAlignmentHelper<S>> = Array((size() + length - 1) / length) { i ->
        val pointer = i * length
        val l = min(length, size() - pointer)
        getRange(pointer, pointer + l)
    }

    class Settings(
        val markMatchWithSpecialLetter: Boolean,
        val lowerCaseMatch: Boolean,
        val lowerCaseMismatch: Boolean,
        val specialMatchChar: Char,
        val outOfRangeChar: Char
    )

    override fun toString(): String = format()

    @JvmOverloads
    fun format(linesFormatter: MultiAlignmentFormatter.LinesFormatter = MultiAlignmentFormatter.LinesFormatter()): String =
        linesFormatter.formatLines(this)

    companion object {
        private fun aabs(pos: Int): Int {
            if (pos >= 0) return pos
            return if (pos == -1) -1 else -2 - pos
        }

        @JvmField
        val DEFAULT_SETTINGS = Settings(
            markMatchWithSpecialLetter = false,
            lowerCaseMatch = true,
            lowerCaseMismatch = false,
            specialMatchChar = ' ',
            outOfRangeChar = ' '
        )
        val DOT_MATCH_SETTINGS = Settings(
            markMatchWithSpecialLetter = true,
            lowerCaseMatch = true,
            lowerCaseMismatch = false,
            specialMatchChar = '.',
            outOfRangeChar = ' '
        )

        @JvmStatic
        @SafeVarargs
        fun <S : Sequence<S>> build(
            settings: Settings,
            subjectRange: Range,
            name: String,
            subject: S,
            inputs: List<Input<S>>,
            metaInfo: List<MetaInfoInput<S>>
        ): MultiAlignmentHelper<S> {
            for (input in inputs) require(input.alignment.sequence1 == subject)
            var subjectPointer = subjectRange.from
            val subjectPointerTo = subjectRange.to
            val aCount = inputs.size
            val queryPointers = IntArray(aCount) { inputs[it].alignment.sequence2Range.from }
            val mutationPointers = IntArray(aCount)
            val mutations: Array<Mutations<S>> = Array(aCount) { inputs[it].alignment.absoluteMutations }
            val matches: Array<MutableList<Boolean>> = Array(aCount) { ArrayList() }
            val subjectPositions = IntArrayList()
            val queryPositions: Array<IntArrayList> = Array(aCount) { IntArrayList() }
            val subjectString = StringBuilder()
            val queryStrings: Array<StringBuilder> = Array(aCount) { StringBuilder() }
            val processed = BitArray(aCount)
            while (true) {
                // Checking continue condition
                var doContinue = subjectPointer < subjectPointerTo
                for (i in 0 until aCount) doContinue = doContinue or (mutationPointers[i] < mutations[i].size())
                if (!doContinue) break
                processed.clearAll()

                // Processing out of range sequences
                for (i in 0 until aCount) {
                    if (!inputs[i].alignment.sequence1Range.contains(subjectPointer)
                        && !(inputs[i].alignment.sequence1Range.containsBoundary(subjectPointer) &&
                                mutationPointers[i] != mutations[i].size())
                    ) {
                        queryStrings[i].append(settings.outOfRangeChar)
                        queryPositions[i].add(-1)
                        matches[i].add(false)
                        processed.set(i)
                    }
                }

                // Checking for insertions
                var insertion = false
                for (i in 0 until aCount) {
                    if (mutationPointers[i] < mutations[i].size() &&
                        mutations[i].getTypeByIndex(mutationPointers[i]) == MutationType.Insertion &&
                        mutations[i].getPositionByIndex(mutationPointers[i]) == subjectPointer
                    ) {
                        insertion = true
                        queryStrings[i].append(mutations[i].getToAsSymbolByIndex(mutationPointers[i]))
                        queryPositions[i].add(queryPointers[i]++)
                        matches[i].add(false)
                        mutationPointers[i]++
                        assert(!processed[i])
                        processed.set(i)
                    }
                }
                if (insertion) { // In case on insertion in query sequence
                    subjectString.append('-')
                    subjectPositions.add(-2 - subjectPointer)
                    for (i in 0 until aCount) {
                        if (!processed[i]) {
                            queryStrings[i].append('-')
                            queryPositions[i].add(-2 - queryPointers[i])
                            matches[i].add(false)
                        }
                    }
                } else { // In other cases
                    val subjectSymbol = subject.symbolAt(subjectPointer)
                    subjectString.append(subjectSymbol)
                    subjectPositions.add(subjectPointer)
                    for (i in 0 until aCount) {
                        if (processed[i]) continue
                        val cMutations = mutations[i]
                        val cMutationPointer = mutationPointers[i]
                        var mutated = false
                        if (cMutationPointer < cMutations.size()) {
                            val mutPosition = cMutations.getPositionByIndex(cMutationPointer)
                            assert(mutPosition >= subjectPointer)
                            mutated = mutPosition == subjectPointer
                        }
                        if (mutated) {
                            when (cMutations.getRawTypeByIndex(cMutationPointer)) {
                                Mutation.RAW_MUTATION_TYPE_SUBSTITUTION -> {
                                    val symbol = cMutations.getToAsSymbolByIndex(cMutationPointer)
                                    queryStrings[i].append(if (settings.lowerCaseMismatch) symbol.lowercaseChar() else symbol)
                                    queryPositions[i].add(queryPointers[i]++)
                                    matches[i].add(false)
                                }
                                Mutation.RAW_MUTATION_TYPE_DELETION -> {
                                    queryStrings[i].append('-')
                                    queryPositions[i].add(-2 - queryPointers[i])
                                    matches[i].add(false)
                                }
                                else -> assert(false)
                            }
                            mutationPointers[i]++
                        } else {
                            queryStrings[i].append(
                                when {
                                    settings.markMatchWithSpecialLetter -> settings.specialMatchChar
                                    settings.lowerCaseMatch -> subjectSymbol.lowercaseChar()
                                    else -> subjectSymbol
                                }
                            )
                            queryPositions[i].add(queryPointers[i]++)
                            matches[i].add(true)
                        }
                    }
                    subjectPointer++
                }
            }
            val matchesBAs: Array<BitArray> = Array(aCount) {
                BitArray(matches[it])
            }
            val queryStringsArray: Array<QueryLine> = Array(aCount) { i ->
                when (val input = inputs[i]) {
                    is AlignmentInput -> AlignmentLine(
                        geneName = input.geneName,
                        content = queryStrings[i].toString(),
                        positions = queryPositions[i].toArray(),
                        alignmentScore = input.alignmentScore,
                        hitScore = input.hitScore
                    )
                    is ReadInput -> ReadLine(
                        index = input.index,
                        content = queryStrings[i].toString(),
                        positions = queryPositions[i].toArray(),
                    )
                }
            }
            return MultiAlignmentHelper(
                SubjectLine(
                    name = name,
                    source = subject,
                    content = subjectString.toString(),
                    positions = subjectPositions.toArray()
                ),
                queryStringsArray,
                matchesBAs,
                metaInfo
            )
        }
    }

    sealed interface MetaInfoInput<S : Sequence<S>>

    class ReferencePointsInput<S : Sequence<S>>(
        val partitioning: SequencePartitioning
    ) : MetaInfoInput<S>

    class AminoAcidInput(
        val partitioning: SequencePartitioning
    ) : MetaInfoInput<NucleotideSequence>

    class QualityInput(
        val quality: SequenceQuality
    ) : MetaInfoInput<NucleotideSequence>

    sealed interface AnnotationLine {
        val content: String
        fun subRange(from: Int, to: Int): AnnotationLine
    }

    data class QualityLine(
        override val content: String
    ) : AnnotationLine {
        override fun subRange(from: Int, to: Int) = copy(
            content = content.substring(from, to)
        )
    }

    data class AminoAcidsLine(
        override val content: String
    ) : AnnotationLine {
        override fun subRange(from: Int, to: Int) = copy(
            content = content.substring(from, to)
        )
    }

    data class ReferencePointsLine(
        override val content: String
    ) : AnnotationLine {
        override fun subRange(from: Int, to: Int) = copy(
            content = content.substring(from, to)
        )
    }

    sealed interface LineWithPositions {
        val content: String
        val positions: IntArray
        fun subRange(from: Int, to: Int): LineWithPositions

        val firstPosition: Int
            get() {
                for (pos in positions) if (pos >= 0) return pos
                for (pos in positions) if (pos < -1) return -2 - pos
                return -1
            }

        val lastPosition: Int
            get() {
                for (i in positions.indices.reversed()) if (positions[i] >= 0) return positions[i]
                for (i in positions.indices.reversed()) if (positions[i] < -1) return -2 - positions[i]
                return -1
            }
    }

    sealed interface QueryLine : LineWithPositions {
        override fun subRange(from: Int, to: Int): QueryLine
    }

    class SubjectLine<S : Sequence<S>>(
        val name: String,
        val source: S,
        override val content: String,
        override val positions: IntArray
    ) : LineWithPositions {
        override fun subRange(from: Int, to: Int): SubjectLine<S> = SubjectLine(
            name = name,
            source = source,
            content = content.substring(from, to),
            positions = positions.copyOfRange(from, to)
        )
    }

    class AlignmentLine(
        val geneName: String,
        override val content: String,
        override val positions: IntArray,
        val alignmentScore: Int,
        val hitScore: Int
    ) : QueryLine {
        override fun subRange(from: Int, to: Int) = AlignmentLine(
            geneName = geneName,
            content = content.substring(from, to),
            positions = positions.copyOfRange(from, to),
            alignmentScore = alignmentScore,
            hitScore = hitScore
        )
    }


    class ReadLine(
        val index: String,
        override val content: String,
        override val positions: IntArray
    ) : QueryLine {
        override fun subRange(from: Int, to: Int) = ReadLine(
            index = index,
            content = content.substring(from, to),
            positions = positions.copyOfRange(from, to)
        )
    }

    sealed interface Input<S : Sequence<S>> {
        val alignment: Alignment<S>
    }

    class AlignmentInput<S : Sequence<S>>(
        val geneName: String,
        override val alignment: Alignment<S>,
        val alignmentScore: Int,
        val hitScore: Int
    ) : Input<S>

    class ReadInput<S : Sequence<S>>(
        val index: String,
        override val alignment: Alignment<S>
    ) : Input<S>
}


// may be used for calculation fo minimalPositionWidth for formatLines
@Suppress("unused")
val Array<MultiAlignmentHelper<*>>.maxPositionWidth: Int
    get() = maxOf { it.actualPositionWidth }

